File: src/main/scala/database/Migrations.scala
package org.jumalauta.partyboi
package database

import cats.effect.Sync
import org.flywaydb.core.Flyway
import org.flywaydb.core.api.Location
import org.flywaydb.core.api.configuration.FluentConfiguration

object Migrations extends Logging {
  private val migrationsLocation = "classpath:migrations"
  def migrate[F[_]: Sync](settings: DatabaseSettings): F[Int] =
    Sync[F].delay {
      log.info(s"Running migrations from classpath:migrations")
      val count = unsafeMigrate(settings)
      log.info(s"Executed $count migrations")
      count
    }

  private def unsafeMigrate(settings: DatabaseSettings): Int = {
    val m = Flyway.configure
      .dataSource(
        settings.url,
        settings.user,
        settings.password,
      )
      .group(true)
      .outOfOrder(false)
      .table("migration")
      .locations(new Location(migrationsLocation))
      .baselineOnMigrate(true)

    m.load().migrate().migrationsExecuted
  }
}

File: src/main/kotlin/party/jml/partyboi/db/Migrations.kt
package party.jml.partyboi.db

import arrow.core.Either
import arrow.core.left
import io.ktor.http.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.flywaydb.core.Flyway
import org.flywaydb.core.api.output.MigrateResult
import org.flywaydb.core.api.output.ValidateOutput
import party.jml.partyboi.data.AppError
import party.jml.partyboi.data.InternalServerError
import party.jml.partyboi.system.AppResult

object Migrations {
    suspend fun migrate(db: DatabasePool, schemaName: String = "public"): AppResult<MigrateResult> =
        withContext(Dispatchers.IO) {
            val config = Flyway.configure()
                .dataSource(db.dataSource)
                .schemas(schemaName)
                .group(true)
                .outOfOrder(false)
                .table("migrations")
                .locations("classpath:db/migrations")
                .validateMigrationNaming(true)
                .baselineOnMigrate(true)

            val validatedConfig = config
                .ignoreMigrationPatterns("*:pending")
                .load()
                .validateWithResult()

            if (validatedConfig.validationSuccessful) {
                Either.catch {
                    config.load().migrate()
                }.mapLeft { InternalServerError(it) }
            } else {
                MigrationError(validatedConfig.invalidMigrations).left()
            }
        }
}

class MigrationError(migrationErrors: List<ValidateOutput>) : AppError {
    override val message: String =
        "Migration failed: ${
            migrationErrors.joinToString(", ") {
                "(version=${it.version}, description=${it.description})"
            }
        }"
    override val statusCode: HttpStatusCode = HttpStatusCode.InternalServerError
    override val throwable: Throwable = Exception(message)
}